src/librustc_typeck/check/_match.rs

```diff
@@ -299,10 +299,19 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                 let element_tys = tcx.mk_type_list(element_tys_iter);
                 let pat_ty = tcx.mk_ty(ty::TyTuple(element_tys));
                 self.demand_eqtype(pat.span, expected, pat_ty);
-                for (i, elem) in elements.iter().enumerate_and_adjust(max_len, ddpos) {
-                    self.check_pat_walk(elem, &element_tys[i], def_bm, true);
+                if self.has_errors.get() {
+                    let element_tys_iter = (0..max_len).map(|_| tcx.types.err);
+                    let element_tys = tcx.mk_type_list(element_tys_iter);
+                    for (_, elem) in elements.iter().enumerate_and_adjust(max_len, ddpos) {
+                        self.check_pat_walk(elem, &tcx.types.err, def_bm, true);
+                    }
+                    tcx.mk_ty(ty::TyTuple(element_tys))
```

Comment:
Can this use `tcx.mk_tup` or something instead? It seems a bit too complex.

Reply:
Fair enough. Will change.

Reply:
I'm just not sure how extensive the effects of `enumerate_and_adjust` and `check_pat_walk` are.

I'll try to get a compile with simplified logic, but will keep it as a separate commit in case it causes uncaught regressions.